Moloco-RMP Destination Integration V1

heonjang opened the pull request:
## Description
This PR implements the V1 of Moloco-RMP Destination.

There will be a total of 8 actions types
1. addToCart
2. addToWishlist
3. home
4. itemPageView
5. land
6. pageView
7. purchase
8. search

## Structure Overview
Moloco expects an almost identical schema for each event type but the requirements differ for each action type.

Therefore, this code tried to modularize as much as possible to minimize redundancy.

Modularized logics are located in the `moloco-rmp/common` directory
### convert.ts
This module contains the conversion logic. It is responsible for converting the given input data into a Moloco-RMP-compatible format.

### fields.ts
This module contains the fields that will be shown in the mapping UI. There are two kinds of fields: Common and Variable fields. The common field will be exposed for all actions identically, but the variable fields can be shown with different requirements as well as their descriptions.

### event.ts
This module will be used by each action to define their details. There, it decides the requirements of each "Variable" field.

### request-client.ts
This module is responsible for sending the data to the Moloco-RMP's end 

## Testing
- [ ✅ ] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality

1. `moloco-rmp/__tests__/convert.test.ts`: Testing the conversion logic.
10. `moloco-rmp/__tests__/index.test.ts`: Action-neutral unit tests.

- Default mappings
- Custom mappings
- Overwriting logics
- Special conversion logic (page identifier tokens)

11. `moloco-rmp/{action}/__tests__/index.test.ts`: Unit tests for each action

- [ ✅ ] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
Done for all action types.
- Tested valid events successfully reached Moloco side
- Tested invalid events threw errors

heonjang commented on packages/destination-actions/src/destinations/moloco-rmp/common/convert.ts:
```diff
@@ -0,0 +1,166 @@
+import { PayloadValidationError } from '@segment/actions-core'
```
This module contains the conversion logic. It is responsible for converting the given input data into a Moloco-RMP-compatible format.

heonjang commented on packages/destination-actions/src/destinations/moloco-rmp/common/fields.ts:
```diff
@@ -0,0 +1,313 @@
+import { InputField } from '@segment/actions-core/destination-kit/types'
```
This module contains the fields that will be shown in the mapping UI. There are two kinds of fields: Common and Variable fields. The common field will be exposed for all actions identically, but the variable fields can be shown with different requirements as well as their descriptions.

heonjang commented on packages/destination-actions/src/destinations/moloco-rmp/common/request-client.ts:
```diff
@@ -0,0 +1,71 @@
+import { 
```
This module is responsible for sending the data to the Moloco-RMP's end

joe-ayoub-segment commented:
Thanks for raising the PR @heonjang - I'll schedule for review.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/moloco-rmp/common/fields.ts:
```diff
@@ -0,0 +1,313 @@
+  choices: [
+    { label: 'App', value: 'APP' },
+    { label: 'Site', value: 'SITE' }
+  ]
```
Would it make sense to set a default?
e.g. 
default: 'APP'

heonjang replied:
I don't think having a default value is a great idea.

Normally, if the integrated source uses ANALYTICS.JS, it would be SITE and APP for ANALYTICS.iOS/Android. Choosing any of them as a default value doesn't seem right

joe-ayoub-segment commented on packages/destination-actions/src/destinations/moloco-rmp/common/fields.ts:
```diff
@@ -0,0 +1,313 @@
+  }
+}
+
+export const DEFAULT_CURRENCY: InputField = {
+  label: 'Default Currency',
+  description: `The default currency value. If this is set, it will be used as a default currency value for items.
+  Available options are the followings
+  UNKNOWN_CURRENCY: Unknown currency.
+  USD: US Dollar.
+  KRW: Korean Won.
+  JPY: Japanese Yen.
+  EUR: EU Euro.
+  GBP: British Pound.
+  SEK: Swedish Krona.
+  INR: India Rupee.
+  THB: Thailand Baht.
+  IDR: Indonesia Rupiah.
+  CNY: China Yuan.
+  CAD: Canada Dollar.
+  RUB: Russia Ruble.
+  BRL: Brazil Real.
+  SGD: Singapore Dollar.
+  HKD: Hong Kong Dollar.
+  AUD: Autrailia Dollar.
+  PLN: Poland Zloty.
+  DKK: Denmark Krone.
+  VND: Viet Nam Dong.
+  MYR: Malaysia Ringgit.
+  PHP: Philippines Peso.
+  TRY: Turkey Lira.
+  VEF: Venezuela Bolívar.`,
+  type: 'string',
+  required: false
+}
```
I recommend offering a list of choices, like this: 

```suggestion
export const DEFAULT_CURRENCY: InputField = {
  label: 'Default Currency',
  description: `The default currency value. If this is set, it will be used as a default currency value for items.',
  choices: [
  {label: 'UNKNOWN_CURRENCY', value: 'UNKNOWN_CURRENCY'},
  {label: 'USD', value: 'USD'},
  {label: 'KRW', value: 'KRW'},
  {label: 'JPY', value: 'JPY'},
  {label: 'EUR', value: 'EUR'},
  {label: 'GBP', value: 'GBP'},
  {label: 'SEK', value: 'SEK'},
  {label: 'INR', value: 'INR'},
  {label: 'THB', value: 'THB'},
  {label: 'IDR', value: 'IDR'},
  {label: 'CNY', value: 'CNY'},
  {label: 'CAD', value: 'CAD'},
  {label: 'RUB', value: 'RUB'},
  {label: 'BRL', value: 'BRL'},
  {label: 'SGD', value: 'SGD'},
  {label: 'HKD', value: 'HKD'},
  {label: 'AUD', value: 'AUD'},
  {label: 'PLN', value: 'PLN'},
  {label: 'DKK', value: 'DKK'},
  {label: 'VND', value: 'VND'},
  {label: 'MYR', value: 'MYR'},
  {label: 'PHP', value: 'PHP'},
  {label: 'TRY', value: 'TRY'},
  {label: 'VEF', value: 'VEF'}
],
  type: 'string',
  required: false
}
```

heonjang replied:
Much readable. Thank you!

joe-ayoub-segment commented on packages/destination-actions/src/destinations/moloco-rmp/common/fields.ts:
```diff
@@ -0,0 +1,313 @@
+      description: 'Currency information'
+      + (required ? '' : ', REQUIRED IF PRICE IS GIVEN)')
+      + ` Available options are the followings
+  UNKNOWN_CURRENCY: Unknown currency.
+  USD: US Dollar.
+  KRW: Korean Won.
+  JPY: Japanese Yen.
+  EUR: EU Euro.
+  GBP: British Pound.
+  SEK: Swedish Krona.
+  INR: India Rupee.
+  THB: Thailand Baht.
+  IDR: Indonesia Rupiah.
+  CNY: China Yuan.
+  CAD: Canada Dollar.
+  RUB: Russia Ruble.
+  BRL: Brazil Real.
+  SGD: Singapore Dollar.
+  HKD: Hong Kong Dollar.
+  AUD: Autrailia Dollar.
+  PLN: Poland Zloty.
+  DKK: Denmark Krone.
+  VND: Viet Nam Dong.
+  MYR: Malaysia Ringgit.
+  PHP: Philippines Peso.
+  TRY: Turkey Lira.
+  VEF: Venezuela Bolívar.`,
```
Same as with Default Currency field above. I'd recommend this be a choice array list.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/moloco-rmp/common/settings.ts:
```diff
@@ -0,0 +1,8 @@
+// Generalized interface for auth.authentication.fields
+// All destination actions, will use the auth.authentication to define the fields for the authentication scheme,
+// and the corresponding interface will be automatically generated.
+// This interface is the generalized version of that auto-generated interface. 
+export interface Settings {
+    platformId: string
+    apiKey: string
+}
```
Not sure why this is needed. Why not reference the actual Settings interface from moloco-rmp/generated-types.ts?

joe-ayoub-segment commented on packages/destination-actions/src/destinations/moloco-rmp/addToCart/index.ts:
```diff
@@ -0,0 +1,24 @@
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Add to Cart',
+  description: 'Represents a user adding an item to his cart',
```
```suggestion
  description: 'Represents a user adding an item to their cart',
```

joe-ayoub-segment commented on packages/destination-actions/src/destinations/moloco-rmp/common/request-client.ts:
```diff
@@ -0,0 +1,71 @@
+    async sendEvent(body: Record<string, any>): Promise<ModifiedResponse> {
+        const res = await this._sendEvent(body)
+
+        // OK
+        if (res.status === 200) {
+            return res
+        } else {
+            const errorMsg = await res.text();
+
+            // Retry on 5xx (server errors) and 429s (rate limits)
+            if (res.status >= 500 || res.status === 429) {
+                throw new RetryableError(`${res.status} | ${errorMsg}, this will be retried`)
+            }
+
+            // Unauthorized errors, could be invalid platform or API Key
+            if (res.status === 401) {
+                throw new InvalidAuthenticationError(`${res.status} | Unauthorized: ${errorMsg}`)
+            }
+
+            // Invalid payload
+            if (res.status === 400) {
+                throw new APIError(`${res.status} | Invalid payload: ${errorMsg}`, res.status)
+            }
+
+            // Else, throw a non-retryable error
+            throw new Error(`${res.status}' | Unknown error: ${errorMsg}`)
+        }
+    }
```
There is no need for any of this code as you are using a RequestClient to send the data (which is the correct thing to do).    

Segment will automatically retry on a 5XX or 429. 
Errors are displayed to the customer in the delivery tab.

joe-ayoub-segment reviewed (changes requested):
Hi @heonjang thanks for raising this PR. 

I spent some time reviewing it and decided to ask that we meet again to discuss how it was put together. 

I believe that it can be simplified a lot. I don't think it will be much work, but it will make the code far easier to understand and maintain. 

Would you be able to schedule a call with me please? Let's go for a 1h meeting if you have the time?

http://calendly.com/joe_ayoub/integration-workshop

Best regards,
Joe

heonjang commented on packages/destination-actions/src/destinations/moloco-rmp/common/fields.ts:
```diff
@@ -0,0 +1,313 @@
+  description: 'Timestamp that the event happened at.',
+  type: 'datetime',
+  required: true,
+  // could set a default value to '$.timestamp' but did not do so
```
Thanks for the suggestion, I will fix so!

heonjang commented on packages/destination-actions/src/destinations/moloco-rmp/common/fields.ts:
```diff
@@ -0,0 +1,313 @@
+
+export const USER_ID: InputField = {
+  label: 'User ID',
+  description: 'User Identifier for the platform. Recommended to hash it before sending for anonymization. The length should not exceed 128 characters.',
```
It was something recommended but not in a strict way. I will remove the recommendation in the description and consider your suggestion as well!

heonjang commented:
@joe-ayoub-segment 

Hi Joe, I've applied the reviews given by you earlier today!

Thank you for your detailed guidance, and please let me know if there is any further improvement to be made.

Best,

Shawn

joe-ayoub-segment commented:
hi @heonjang - would you be available to catch up before the deploy on Tuesday please?

I refactored the code and added some stuff which I think you might want to include. 

I'm available for the next few hours if you are free?

joe-ayoub-segment commented:
Hi @heonjang I raised a separate PR with some recommended changes. https://github.com/segmentio/action-destinations/pull/1917 

I tried to push the commit to a branch on your repo but I didn't have permission. 

Here's a [video](https://drive.google.com/file/d/1ezFARkN5SwFwriNrtpMrfTHX6BRWozsR/view?usp=sharing) of me walking through the changes.

heonjang commented:
> hi @heonjang - would you be available to catch up before the deploy on Tuesday please?
> 
> I refactored the code and added some stuff which I think you might want to include.
> 
> I'm available for the next few hours if you are free?

Sorry, I couldn't make it this morning. I will schedule it on Monday morning through Calendly!

heonjang commented:
> Hi @heonjang I raised a separate PR with some recommended changes. #1917
> 
> I tried to push the commit to a branch on your repo but I didn't have permission.
> 
> Here's a [video](https://drive.google.com/file/d/1ezFARkN5SwFwriNrtpMrfTHX6BRWozsR/view?usp=sharing) of me walking through the changes.

Thank you so much. I will watch the video and apply the updates you suggested

joe-ayoub-segment commented:
Closing PR as replaced by https://github.com/segmentio/action-destinations/pull/1903
